### PR TITLE
This is for issue #1712

### PR DESCRIPTION
**Pull request by krethan**
This is for issue #1712

Simplify the logic in the IBMi class by creating new IBMiSettings class.

### CODE REVIEW

**Comment by worksofliam** [comment]
@krethan This is a nice change. I am going to work with the team to get a formal review ready - but in the meantime, I think the built-in formatter should be run over the `IBMiSettings.ts` file. You can open the command palette while inside that file, and search for 'Format Document'.

Thanks!

**Comment by krethan** [comment]
@worksofliam Formatted IBMi.ts and IBMiSettings.ts

**Comment by krethan** [comment]
@worksofliam Fixed merge issue in case it was preventing review.

**Comment by worksofliam** [comment]
@krethan just wanted you to know that we are planning on taking a stab at reviewing this next week. We have been quite busy! Sorry about the wait.

File: src/api/IBMiSettings.ts [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+    names: [`QZDFMDB2.PGM`]
+  },
+  {
+    path: `/QIBM/ProdData/IBMiDebugService/bin/`,
+    specific: `startDebugService.sh`,
+    names: [`startDebugService.sh`]
+  }
+];
+
+export default class IBMiSettings {
```
Comment (worksofliam) [inline src/api/IBMiSettings.ts:34]: @sebjulliand and I were chatting about this class name and I think something more like `ConnectionSettings` or `InstanceSettings` might be more appropriate.

File: src/api/IBMiSettings.ts [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+
+  private connection: IBMi;
+  private progress: vscode.Progress<{ message?: string; increment?: number }>;
+  private reconnecting?: boolean;
+  private connectionObject: ConnectionData;
+  private tempLibrarySet: boolean;
+  private tempDirSet: boolean;
+  private delayedOperations: Function[];
+
+  constructor(connection: IBMi, progress: vscode.Progress<{ message?: string; increment?: number }>, connectionObject: ConnectionData, delayedOperations: Function[], reconnecting?: boolean) {
```
Comment (worksofliam) [inline src/api/IBMiSettings.ts:44]: This constructor is likely too long considering most of the parameters end up as private member fields.

Please consider using `private connection: IBMi, private progress: ...`, etc, which then will mean you don't need to assign in the constructor or define them above the constructor.

Comment (worksofliam) [reply src/api/IBMiSettings.ts:44]: More detail here: https://stackoverflow.com/a/48571668

**Review by worksofliam — CHANGES_REQUESTED** [review]
Generally on the right track. I've left some initial comments, but I believe there is likely a more to do outside of my review.

I don't know if the rest of @codefori/core agree with me, but ideally, I'd like to see this settings class not make use of the `vscode` namespace at all, and instead return the suffice information from each class method so the UI can be handled independently.

File: src/api/IBMiSettings.ts (hunk 1 above)
Comment (krethan) [reply src/api/IBMiSettings.ts:34]: @worksofliam I agree we should seperate the UI from this class. I will refactor the UI code back into the connection method of the IBMi class.

Comment (krethan) [reply src/api/IBMiSettings.ts:34]: @worksofliam @sebjulliand I will go with ConnectionSettings for the class name

**Comment by krethan** [comment]
@worksofliam I am about 75% through refactoring the ConnectionSettings.ts class now.

**Comment by krethan** [comment]
@worksofliam Have couple introduced errors now. Working my way through them. Bug in home directory and another when asked to set bash to default, disconnect, connect again and still asking to set bash. Will ask for code review after these are fixed and committed.

**Comment by krethan** [comment]
@worksofliam Okay ready for new review. I hope this one is good.

**Comment by worksofliam** [comment]
@krethan As we get ready for the 2.8.0 release, is it possible you can fix the conflicts here? Thanks!

**Comment by worksofliam** [comment]
@krethan It looks like the entire `out` directory has been committed. This likely shouldn't be the case. Any chance you can take a look at that?

**Comment by krethan** [comment]
@worksofliam I reverted the commits back to remove everything in the out directory. Makes sense.

**Comment by worksofliam** [comment]
@krethan We are so close! Please could you also fix the last remaining conflict?

Sorry this is taking a while. This is a very high impact PR so we are treading lightly right now.

File: src/api/ConnectionSettings.ts [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+          if (badLib) badLibs.push(badLib);
+        });
+      }
+    }
+    
+    return badLibs;
+
+  }
+
+  private appendOutput(content: string) {
```
Comment (worksofliam) [inline src/api/ConnectionSettings.ts:557]: Looks like this method could be removed as it is unreferenced.

File: src/api/ConnectionSettings.ts [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+      command: `CHKOBJ OBJ(QSYS/QCPTOIMPF) OBJTYPE(*DTAARA)`,
+      noLibList: true
+    });
+    if (QCPTOIMPF?.code === 0) {
+      dataArea = true;
+    }
+    return dataArea;
+  }
+
+  async deleteQCPTOIMPF(): Promise<CommandResult>{
```
Comment (worksofliam) [inline src/api/ConnectionSettings.ts:295]: This function is returning a function, which returns a Promise, so there is no need for the `async` keyword on the method here.

File: src/api/ConnectionSettings.ts [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+    });
+
+    if (QCPFRMIMPF?.code === 0) {
+      dataArea = true;
+    }
+
+    return dataArea;
+  }
+
+  async deleteQCPFRMIMPF(): Promise<CommandResult> {
```
Comment (worksofliam) [inline src/api/ConnectionSettings.ts:318]: Same here with the `async` keyword.

File: src/api/ConnectionSettings.ts [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+            if (!messages.findId(`CPF2125`)) {
+              // @ts-ignore We know the config exists.
+              throw(`Temporary data not cleared from ${this.config.tempLibrary}.`);
+            }
+          }
+        })
+    }
+  }
+
+  async clearTempDataIFS() {
```
Comment (worksofliam) [inline src/api/ConnectionSettings.ts:265]: `async` keyword is not required as `await` is not being used in this method.

File: src/api/ConnectionSettings.ts [hunk 7]
```diff
@@ -0,0 +1,563 @@
...
+        // Directory not created
+      }
+    }
+
+    if (!this.tempDirSet && this.connection.config) {
+      this.connection.config.tempDir = `/tmp`;
+    }
+  }
+
+  async clearTempDataSys(){
```
Comment (worksofliam) [inline src/api/ConnectionSettings.ts:246]: `async` keyword not required here either.

File: src/api/IBMi.ts [hunk 8]
```diff
@@ -210,274 +183,112 @@ export default class IBMi {
...
-          } else {
-            // Directory not created
-          }
-        }
-
-        if (!tempDirSet) {
-          this.config.tempDir = `/tmp`;
-        }
+        //Clear temporary data
+        if (await connSettings.getTempLibrarySet() && this.config?.autoClearTempData) {
```
Comment (worksofliam) [inline src/api/IBMi.ts:245]: `await` not required here because `getTempLibrarySet` does not return a Promise.

File: src/api/IBMi.ts [hunk 9]
```diff
@@ -486,35 +297,29 @@ export default class IBMi {
             });
           }
 
-          const QCPFRMIMPF = await this.runCommand({
-            command: `CHKOBJ OBJ(QSYS/QCPFRMIMPF) OBJTYPE(*DTAARA)`,
-            noLibList: true
-          });
-
-          if (QCPFRMIMPF?.code === 0) {
+          if (await connSettings.checkQCPFRMIMPF) {
```
Comment (worksofliam) [inline src/api/IBMi.ts:300]: Brackets are missing from this method call.

File: src/api/IBMi.ts [hunk 10]
```diff
@@ -526,39 +331,14 @@ export default class IBMi {
...
-                } else {
-                  for (const name of feature.names)
-                    if (files.includes(name))
-                      this.remoteFeatures[name] = feature.path + name;
-                }
-              }
-            } catch (e) {
-              console.log(e);
-            }
+          const remoteApps = await connSettings.getRemoteApps();
```
Comment (worksofliam) [inline src/api/IBMi.ts:336]: `await` not required since `getRemoteApps` does not return a Promise.

**Review by worksofliam — CHANGES_REQUESTED** [review]
I have left you a review to get you started. There are lot of ends that need some cleaning up, espcially in regards to the use of `async` and `await`. I have high hopes for this PR, but it needs to be tight and tested before we can merge.

Other suggestions:

* Use the VS Code built-in formatter on `ConnectionSettings.ts`: F1 -> Format Document
* A few methods in `ConnectionSettings` use try catch, even though the calls inside of them do not crash - for example, `sendCommand` does not crash, but returns the exit code. It'd be nice to have those removed.
* A lot of the methods return `void`. It would be nice to make them more verbose, perhaps returning `boolean` or other in the future so we can write test cases for them.

File: src/api/ConnectionSettings.ts [hunk 11]
```diff
@@ -0,0 +1,563 @@
...
+          "export PATH=/QOpenSys/pkgs/bin:$PATH"
+        );
+      }
+      await content.writeStreamfile(bashrcFile, bashrcContent.join("\n"));
+    }
+  }
+
+  async validateLibraryList(): Promise<string[]> {
+    
+    let validLibs: string[] = [];
```
[the remark this answers is not in the document]
Comment (krethan) [reply src/api/ConnectionSettings.ts:530]: Removed.

File: src/api/ConnectionSettings.ts (hunk 3 above)
Comment (krethan) [reply src/api/ConnectionSettings.ts:557]: Method removed.

File: src/api/ConnectionSettings.ts (hunk 4 above)
Comment (krethan) [reply src/api/ConnectionSettings.ts:295]: removed Async.

**Comment by krethan** [comment]
@worksofliam Should be good for review again.

**Pull request closed without merging**